### ISSUE TITLE
pad_missing: True is broken for bin edges with many significant figures 

### ISSUE THREAD

**Issue by rob-tay** [FAST-HEP/fast-carpenter#43]
Exception raised in call to `pd.reindex` in `densify_dataframe`. Seems to be something to do with the number of significant figures in bin edges as it can be fixed by `np.round(bins,3)`

https://github.com/FAST-HEP/fast-carpenter/blob/0a2a1dd75bd6ac248931683f854f067211a3917f/fast_carpenter/summary/binned_dataframe.py#L79-L81

```bash
Traceback (most recent call last):                                                                                                                                                            
  File "/global/homes/r/rtaylor/.conda/envs/rabit-dev/bin/fast_carpenter", line 11, in <module>                                                                                               
    load_entry_point('fast-carpenter', 'console_scripts', 'fast_carpenter')()                                                                                                                 
  File "/global/u1/r/rtaylor/RaBiT/sim/fast-carpenter/fast_carpenter/__main__.py", line 71, in main                                                                                           
    _, ret_val = run_carpenter(sequence, datasets, args)                                                                                                                                      
  File "/global/u1/r/rtaylor/RaBiT/sim/fast-carpenter/fast_carpenter/__main__.py", line 88, in run_carpenter
    ret_val = process.run(datasets, sequence)                                                                                                                                         [0/1910]
  File "/global/homes/r/rtaylor/.conda/envs/rabit-dev/lib/python3.6/site-packages/atuproot/atuproot_main.py", line 59, in run
    result = self._run(loop)
  File "/global/homes/r/rtaylor/.conda/envs/rabit-dev/lib/python3.6/site-packages/atuproot/atuproot_main.py", line 121, in _run
    result = loop()
  File "/global/homes/r/rtaylor/.conda/envs/rabit-dev/lib/python3.6/site-packages/alphatwirl/datasetloop/loop.py", line 31, in __call__
    return self.reader.end()
  File "/global/homes/r/rtaylor/.conda/envs/rabit-dev/lib/python3.6/site-packages/alphatwirl/datasetloop/reader.py", line 30, in end
    return [reader.end() for reader in self.readers]
  File "/global/homes/r/rtaylor/.conda/envs/rabit-dev/lib/python3.6/site-packages/alphatwirl/datasetloop/reader.py", line 30, in <listcomp>
    return [reader.end() for reader in self.readers]
  File "/global/homes/r/rtaylor/.conda/envs/rabit-dev/lib/python3.6/site-packages/alphatwirl/loop/EventDatasetReader.py", line 96, in end
    return self.collector.collect(dataset_readers_list)
  File "/global/homes/r/rtaylor/.conda/envs/rabit-dev/lib/python3.6/site-packages/atuproot/CollectorComposite.py", line 15, in collect
    for dataset, readerComposites in dataset_readers_list
  File "/global/u1/r/rtaylor/RaBiT/sim/fast-carpenter/fast_carpenter/summary/binned_dataframe.py", line 20, in collect
    output = self._prepare_output(dataset_readers_list)
  File "/global/u1/r/rtaylor/RaBiT/sim/fast-carpenter/fast_carpenter/summary/binned_dataframe.py", line 26, in _prepare_output
    binnings=self.binnings)
  File "/global/u1/r/rtaylor/RaBiT/sim/fast-carpenter/fast_carpenter/summary/binned_dataframe.py", line 40, in combined_dataframes
    output = densify_dataframe(output, binnings)
  File "/global/u1/r/rtaylor/RaBiT/sim/fast-carpenter/fast_carpenter/summary/binned_dataframe.py", line 82, in densify_dataframe
    out_df = in_df.reindex(index=out_index, copy=False)
  File "/global/homes/r/rtaylor/.conda/envs/rabit-dev/lib/python3.6/site-packages/pandas/util/_decorators.py", line 197, in wrapper
    return func(*args, **kwargs)
  File "/global/homes/r/rtaylor/.conda/envs/rabit-dev/lib/python3.6/site-packages/pandas/core/frame.py", line 3809, in reindex
    return super(DataFrame, self).reindex(**kwargs)
  File "/global/homes/r/rtaylor/.conda/envs/rabit-dev/lib/python3.6/site-packages/pandas/core/generic.py", line 4356, in reindex
    fill_value, copy).__finalize__(self)
  File "/global/homes/r/rtaylor/.conda/envs/rabit-dev/lib/python3.6/site-packages/pandas/core/frame.py", line 3741, in _reindex_axes
    fill_value, limit, tolerance)
  File "/global/homes/r/rtaylor/.conda/envs/rabit-dev/lib/python3.6/site-packages/pandas/core/frame.py", line 3749, in _reindex_index
    tolerance=tolerance)
  File "/global/homes/r/rtaylor/.conda/envs/rabit-dev/lib/python3.6/site-packages/pandas/core/indexes/multi.py", line 2225, in reindex
    tolerance=tolerance)
  File "/global/homes/r/rtaylor/.conda/envs/rabit-dev/lib/python3.6/site-packages/pandas/core/indexes/multi.py", line 2174, in get_indexer
    indexer = self._engine.get_indexer(target)
  File "pandas/_libs/index.pyx", line 645, in pandas._libs.index.BaseMultiIndexCodesEngine.get_indexer
  File "pandas/_libs/index.pyx", line 639, in pandas._libs.index.BaseMultiIndexCodesEngine._extract_level_codes
  File "/global/homes/r/rtaylor/.conda/envs/rabit-dev/lib/python3.6/site-packages/pandas/core/indexes/interval.py", line 833, in get_indexer
    indexer = self._get_reindexer(target)
  File "/global/homes/r/rtaylor/.conda/envs/rabit-dev/lib/python3.6/site-packages/pandas/core/indexes/interval.py", line 887, in _get_reindexer
    value = self[i]
  File "/global/homes/r/rtaylor/.conda/envs/rabit-dev/lib/python3.6/site-packages/pandas/core/indexes/interval.py", line 1003, in __getitem__
    result = self._data[value]
  File "/global/homes/r/rtaylor/.conda/envs/rabit-dev/lib/python3.6/site-packages/pandas/core/arrays/interval.py", line 467, in __getitem__
    left = self.left[value]
  File "/global/homes/r/rtaylor/.conda/envs/rabit-dev/lib/python3.6/site-packages/pandas/core/indexes/base.py", line 3958, in __getitem__
    return getitem(key)
IndexError: index 40 is out of bounds for axis 0 with size 34

**Comment by simonecid** [comment]
```pad_missing: True``` seems always leave empty bins, using a scribbler like this:
```yaml
DeltaPtOverPtBinner:
 dataset_col: True
 pad_missing: True
 binning:
     - {in: genJetPt, out: genJetPt, bins: {edges:[20, 30, 40, 50, 60, 70, 80, 100, 125, 150, 175, 200, 250]}}
     - {in: absGenJetEta, out: absGenJetEta, bins: {edges: [0, 1.4, 2.4, 2.964, 5]} }
     - {in: deltaPtOverPt, out: deltaPtOverPt, bins: {low: -1, high: 1, nbins: 50}}     
     - {in: caloJetPileup, out: caloJetPileup, bins: {low: 0, high: 400, nbins: 40}}```
```

The same scribbler produces bins with content when ```pad_missing: False```.

A populated bin with ```pad_missing: False``` is:
```yaml
Calibrated_CaloJet_PU200_Puppi_PFCandidates,"[-inf, 20.0)","[0.0, 1.4)","[1.0, inf)","[0.0, 10.0)"
```
with ```pad_missing: True``` the same bin becomes:
```yaml
Calibrated_CaloJet_PU200_Puppi_PFCandidates,"[-inf, 20.0)","[0.0, 1.399999976158142)","[1.0, inf)","[0.0, 10.0)"
```

and it is empty. Could it be connected to this issue?

**Comment by benkrikler** [comment]
I believe this has finally been addressed in PR #116 so I'm going to close this issue.
